feat: add markdown support for user notes

jacob-thompson opened the pull request:
<!--
    Please include a summary of what you've changed and what issue is fixed.
    In the case of a bug fix, please include steps to reproduce the bug so the pull request can be tested.
    If this PR fixes an issue on GitHub, mention this here to automatically close it: "Fixes #1234.".
-->

fixes #6319

extends the `Label` widget to render markdown for user notes, and adds a toggle for a live markdown preview to the `EditUserNotesDialog`

<img width="363" height="231" alt="image" src="https://github.com/user-attachments/assets/bd7e5c3a-a04d-4a91-920a-219f4dfb4506" />

https://github.com/user-attachments/assets/817a31b5-591f-4750-a455-b65f03ac5ce2

fourtf commented:
Hello Jacob, thank you for your submission. It looks technically sound and I like the usage of QTextDocument to handle markdown.

What I fear is that this feature will only be available to very technical users because it's necessary to know markdown to use it. I also fear that the markdown preview is confusing for users as it will likely show the same text as the left side if no markup is added.

To open this feature up for everyone I think it's necessary to hide the fact that we are using markdown under the hood. A graphical way of formatting (a toolbar with buttons for bold, italics, lists, etc.) would be intuitive and inviting. After all, this is a sub-dialog so why not keep the user interface simple there. The document can be loaded and saved as markdown automatically, so it would "just work" for the user.

What do you think about this? It's also possible to go ahead with your suggested GUI for now and reconsider it later.

Mm2PL commented:
I don't think we need to hide that we're using markdown. I think we 
should take some inspiration from Discord and don't do full WYSIWYG but 
instead do buttons for inserting md syntax and syntax highlighting like:
  - <b>**this text is bold**</b>
  - <em>_this is in italics_</em>
  - ...


W dniu 26.09.2025 o 13:23, fourtf pisze:
> *fourtf* left a comment (Chatterino/chatterino2#6490) <https:// 
> github.com/Chatterino/chatterino2/pull/6490#issuecomment-3338215817>
> 
> Hello Jacob, thank you for your submission. It looks technically sound 
> and I like the usage of QTextDocument to handle markdown.
> 
> What I fear is that this feature will only be available to very 
> technical users because it's necessary to know markdown to use it. I 
> also fear that the markdown preview is confusing for users as it will 
> likely show the same text as the left side if no markup is added.
> 
> To open this feature up for everyone I think it's necessary to hide the 
> fact that we are using markdown under the hood. A graphical way of 
> formatting (a toolbar with buttons for bold, italics, lists, etc.) would 
> be intuitive and inviting. After all, this is a sub-dialog so why not 
> keep the user interface simple there. The document can be loaded and 
> saved as markdown automatically, so it would "just work" for the user.
> 
> What do you think about this? It's also possible to go ahead with your 
> suggested GUI for now and reconsider it later.

fourtf commented:
Yeah maybe it's fine. The user doesn't have to use the markdown preview if he doesn't understand or want it.

jacob-thompson commented:
<img width="689" height="67" alt="image" src="https://github.com/user-attachments/assets/3e0e7b28-82b1-42dc-ab82-30d81c367917" />

pajlada commented on src/widgets/Label.hpp:
```diff
@@ -5,7 +5,10 @@
 
 #include <pajlada/signals/signalholder.hpp>
 
+#include <memory>
+
 class QFontMetricsF;
+class QTextDocument;
```
You can undo the include & forward-declaration of QTextDocument here.

pajlada commented on src/widgets/MarkdownLabel.hpp:
```diff
@@ -0,0 +1,37 @@
+class MarkdownLabel : public Label
+{
+public:
+    explicit MarkdownLabel(QString text = QString(),
+                           FontStyle style = FontStyle::UiMedium);
+    explicit MarkdownLabel(BaseWidget *parent, QString text = QString(),
+                           FontStyle style = FontStyle::UiMedium);
```
Seems like we only use the default-constructor. I'd prefer if you only create one constructor that requires the parent & default text, meaning the places where you create this with the LayoutCreator emplace would need to change a bit.

so something like
```c++
MarkdownLabel(BaseWidget *parent, QString text, FontStyle style = ....)
```

pajlada commented on src/widgets/MarkdownLabel.hpp:
```diff
@@ -0,0 +1,37 @@
+    void updateSize() override;
+    void initializeMarkdownDocument();
+
+    mutable std::unique_ptr<QTextDocument> markdownDocument_;
```
This shouldn't need to be mutable anymore

pajlada commented on resources/buttons/bold-darkMode.svg:
👍 Preferably use icons from https://fluenticons.co/outlined instead where possible, and where not possible let us know where you got these icons from so we can correctly attribute them / follow their license.

pajlada reviewed (changes requested):
Layout nits:
I think the "markdown preview" checkbox should be on the right side of the top bar

The style buttons should probably be grouped a little closer - right now they stretch to take up as much space as they can

Not a full review, but overall seems cool

Thank you!

pajlada reviewed (changes requested):
Overall looks & works well enough. Thanks for your work on this feature.
I pushed some minor fixes, let me know if any of those changes look off to you.

Extra non-blocking feedback, let me know if you want to work on improving its UX or if you want to leave that to future PRs:
 - The heading and list button should probably insert the text at the start of the selected line, not at the cursor position.
 - The heading button works a bit weird with selected text. Maybe improved after the previous nit is fixed?
 - The bold, italizice, and link button inserts a newline at the end _if_ you have a selection. This is ignored in markdown, but still looks weird in the raw text.
 - Clicking the heading/bold/italizice button again should, imo, try to remove the boldness/italicness/heading of the word/line if it's currently active. Very stretch goaly though.

jacob-thompson commented:
> * The heading and list button should probably insert the text at the start of the selected line, not at the cursor position.

in the case of the heading button, I thought the same early on, but I opted to copy the behavior of GitHub's markdown editor

in the case of the list button, you are right and I've just fixed this

> * The bold, italizice, and link button inserts a newline at the end _if_ you have a selection. This is ignored in markdown, 
but still looks weird in the raw text.

this was also a mistake and I've just fixed this
